geth/transactions/addrlock.go

```diff
@@ -27,14 +29,16 @@ func (l *AddrLocker) lock(address common.Address) *sync.Mutex {
 	return l.locks[address]
 }
 
-// LockAddr locks an account's mutex. This is used to prevent another tx getting the
+// LockNonceFor locks an account's mutex and returns a nonce. This is used to prevent another tx getting the
 // same nonce until the lock is released. The mutex prevents the (an identical nonce) from
 // being read again during the time that the first transaction is being signed.
-func (l *AddrLocker) LockAddr(address common.Address) {
+func (l *AddrLocker) LockNonceFor(address common.Address) uint64 {
 	l.lock(address).Lock()
+	return l.localNonce[address]
 }
 
-// UnlockAddr unlocks the mutex of the given account.
-func (l *AddrLocker) UnlockAddr(address common.Address) {
+// UnlockNonceFor unlocks the mutex of the given account and updates nonce.
+func (l *AddrLocker) UnlockNonceFor(address common.Address, nonce uint64) {
+	l.localNonce[address] = nonce
```

Comment:
Race condition?

Reply:
not necessarily, it depends on the usage

Reply:
So like an ordinary race condition :D `AddrLocker` should be immune to bad usage in my opinion. Ideally, it could verify if a given `address` is actually locked. If not, an error should be returned.

Reply:
verifying locked address in golang is not that easy, it requires one more synchronization mechanism like Atomic.CompareAndSwap